modules/imgcodecs/src/grfmt_jpegxl.cpp

```diff
@@ -0,0 +1,376 @@
+
+ImageDecoder JpegXLDecoder::newDecoder() const
+{
+    return makePtr<JpegXLDecoder>();
+}
+
+bool JpegXLDecoder::read(Mat* pimg)
+{
+    // Open file
+    if (!m_f) {
+        m_f.reset(fopen(m_filename.c_str(), "rb"));
+        if (!m_f)
+            return false;
+    }
+
+    // Initialize decoder
+    if (!m_decoder) {
+        m_decoder = JxlDecoderMake(nullptr);
+        if (!m_decoder)
+            return false;
+        // Subscribe to the basic info event
+        JxlDecoderStatus status = JxlDecoderSubscribeEvents(m_decoder.get(), JXL_DEC_BASIC_INFO | JXL_DEC_FULL_IMAGE);
+        if (status != JXL_DEC_SUCCESS)
+            return false;
+    }
+
+    // Set up parallel m_parallel_runner
+    if (!m_parallel_runner) {
+        m_parallel_runner = JxlThreadParallelRunnerMake(nullptr, JxlThreadParallelRunnerDefaultNumWorkerThreads());
+        if (JXL_DEC_SUCCESS != JxlDecoderSetParallelRunner(m_decoder.get(),
+                                                            JxlThreadParallelRunner,
+                                                            m_parallel_runner.get())) {
+            return false;
+        }
+    }
+
+    // Create buffer for reading
+    const size_t read_buffer_size = 16384;  // 16KB chunks
+    if (m_read_buffer.capacity() < read_buffer_size)
+        m_read_buffer.resize(read_buffer_size);
+
+    // Create image if needed
+    if (m_type != -1 && pimg) {
+        pimg->create(m_height, m_width, m_type);
+        if (!pimg->isContinuous())
+            return false;
+        if (JXL_DEC_SUCCESS != JxlDecoderSetImageOutBuffer(m_decoder.get(), 
+                                                        &m_format,
+                                                        pimg->ptr<uint8_t>(),
+                                                        pimg->total() * pimg->elemSize())) {
+            return false;
+        }
+    }
+
+    // Start decoding loop
+    JxlDecoderStatus status = JXL_DEC_NEED_MORE_INPUT;
+    do {
+        // Check if we need more input
+        if (status == JXL_DEC_NEED_MORE_INPUT) {
+            size_t remaining = JxlDecoderReleaseInput(m_decoder.get());
+            // Move any remaining bytes to the beginning
+            if (remaining > 0) {
+                memmove(m_read_buffer.data(), m_read_buffer.data() + m_read_buffer.size() - remaining, remaining);
+            }
+            
+            // Read more data from file
+            size_t bytes_read = fread(m_read_buffer.data() + remaining, 
+                                    1, m_read_buffer.size() - remaining, m_f.get());
+            if (bytes_read == 0) {
+                if (ferror(m_f.get())) {
+                    throw std::runtime_error("Error reading input file");
+                }
+                // If we reached EOF but decoder needs more input, file is truncated
+                if (status == JXL_DEC_NEED_MORE_INPUT) {
+                    throw std::runtime_error("Truncated JXL file");
+                }
```

Comment:
OpenCV's `imread` is exception safe. Please use CV_LOG_WARNING and analogs and return satus instead of exception.

Reply:
done